### PR TITLE
Add 01-basics/

### PR DESCRIPTION
**Pull request by Adrianzatreanu**

### CODE REVIEW

File: 01-basics/Basics.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  Test obiect = new Test(); // obiect este o instanta a clasei Test
+
+Metode vs functii
+-----------------
+
+Functiile sunt exact ca cele din alte limbaje. De aici incolo voi folosi
+notiunea de metoda: o metoda este o functie inauntrul unei clase:
+
+public class Test{
+  void printX(){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: Leave a space between printX() and {

File: 01-basics/Basics.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    System.out.println("x"); // echivalentul a "cout << x;" din C++
```
Comment (razvanch) [inline 01-basics/Basics.rst]: cout << "x";

File: 01-basics/Basics.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+Exemplu instanta:
+  Test obiect = new Test(); // obiect este o instanta a clasei Test
+
+Metode vs functii
+-----------------
+
+Functiile sunt exact ca cele din alte limbaje. De aici incolo voi folosi
+notiunea de metoda: o metoda este o functie inauntrul unei clase:
```
Comment (razvanch) [inline 01-basics/Basics.rst]: O metoda e mai degraba o functie care opereaza pe un tip de date, o clasa definita de tine.

File: 01-basics/Basics.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+Aici printX este o metoda, deoarece este in interiorul clasei Test.
+
+Constructor
+-----------
+
+Constructorul este o metoda care se apeleaza la instantierea unui obiect
+(la apelarea keywordului new). Daca nu definim noi unul, compilatorul ne va
+face unul default, care initializeaza toate campurile cu valorile lor default
+(pentru int/float este 0, pentru String este NULL, etc).
```
Comment (razvanch) [inline 01-basics/Basics.rst]: 'null' instead of NULL.

File: 01-basics/Basics.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Pentru a defini un constructor, nu ii vom pune tip de return si acesta va
+avea numele clasei:
+
+public class Test{
+  public Test(){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: Space before {

File: 01-basics/Basics.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+Astfel, la crearea unui nou obiect in main:
+  Test newObject = new Test(); // prints "Instantiem un nou obiect"
+Practic, in interiorul unui constructor ne putem defini orice fel de comportament,
+pana aici importanta este constientizarea existentei acestuia.
+
+Keywordul "this"
+----------------
+
+Keywordul "this" este o referinta care se refera la instanta curenta. Cu alte
```
Comment (razvanch) [inline 01-basics/Basics.rst]: o referinta care se refera la instanta curenta -> o referinta la instanta curenta

File: 01-basics/Basics.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+cuvinte, daca am avea si un "x" camp in clasa, si un "x" parametru in functie:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  private int x;
+
+  public metoda(int x){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    System.out.println(x); // afiseaza pe "x" parametrul functiei
+    System.out.println(this->x); // afiseaza pe "x" dinauntrul clasei
+  }
+}
+
+Metoda main
+-----------
+
+Pentru a rula codul, in Java efectiv rulam o clasa (click-dreapta pe un fisier
```
Comment (razvanch) [inline 01-basics/Basics.rst:78]: Ar trebui mentionat ca doar clasele cu metoda main definita pot fi rulate.

File: 01-basics/Basics.rst [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+si apoi Run) si acesta va apela metoda main din interiorul clasei sale.
+
+Semnatura este: public static void main(String[] args) {}. Un cod complet
+cu noile notiuni ar fi:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  private int x;
+  private int y;
+
+  public Test(int x, int y){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    this.x = x; // setam x si y din clasa sa fie cei dati ca parametru
+    this.y = y;
+  }
+
+  public void showXY(){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Completari.rst [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+==========
+
+Supraincarcarea constructorului
+-------------------------------
+
+O notiune foarte importanta este cea de supraincarcare: posibilitatea
+de a avea mai multe metode cu acelasi nume dar tip/numar de parametri diferit.
+Momentan ne va interesa doar supraincarcarea constructorului.
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+  private int x;
+  private int y;
+
+  // 1
+  public Test(int x, int y){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    this.x = x;
+    this.y = y;
+  }
+
+  // 2
+  public Test(){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    this.x = 0;
+    this.y = 0;
+  }
+
+  public void showXY(){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    System.out.println("x este: " + this.x + ", iar y: " + this.y); // "+" concateneaza doua stringuri
+  }
+
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+Este important de notat ca fiecare instanta are zona sa de memorie, astfel
+campurile din "a" nefiind in nicio legatura cu cele din "b".
+Astfel, modificand orice camp din "a", nu le vom afecta pe cele din "b".
+
+Specificatori de acces, getteri si setteri
+------------------------------------------
+
+In Java, absolut orice (camp, metoda) trebuie sa aiba un specificator de acces.
+Acesta arata de unde poate fi accesat campul::
+public - permite acces complet din exteriorul clasei curente
```
Comment (razvanch) [inline 01-basics/Completari.rst]: atat din exterior cat si din interior

File: 01-basics/Completari.rst [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+private - limiteaza accesul doar în cadrul clasei curente
+protected - limiteaza accesul doar în cadrul clasei curente si al tuturor descendentilor ei
+      (conceptul de descendenta sau de mostenire va fi explicat mai tarziu)
+(default) - in cazul în care nu este utilizat explicit nici unul din specificatorii
+    de acces de mai sus, accesul este permis doar în cadrul pachetului (package private).
+
+Exemplu:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+  public int x;
+}
+
+public class OtherClass{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    Test a = new Test();
+    a.x = 7;
+    System.out.println(a.x); // prints 7
+  }
+}
+
+VS:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+  private int x;
+}
+
+public class OtherClass{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+    Test a = new Test();
+    a.x = 7; // eroare, in clasa OtherClass nu putem vedea ca Test are un membru x
+    System.out.println(a.x);
+  }
+}
+
+Cum rezolvam?
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+  private int x;
+
+  public void setX(int x){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+    this.x = x;
+  }
+
+  public int getX(){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+    return this.x;
+  }
+}
+
+public class OtherClass{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+Metoda toString
+---------------
+
+Observam ca daca vrem sa afisam o instanta a clasei Test cu println, ne va afisa
+zona de memorie a acelei instante. Cum evitam acest lucru? Suprascriind metoda
+toString (tineti minte semnatura):
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+  private int x;
+  private int y;
+
+  public Test(int x, int y){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+    this.x = x;
+    this.y = y;
+  }
+
+  public String toString(){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/Completari.rst [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+    return "x este: " + this.x + ", iar y: " + this.y;
+  }
+
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Completari.rst]: space before {

File: 01-basics/src/Complex.java [hunk 34]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Created by nfswo on 25-Mar-17.
+ */
+public class Complex {
```
Comment (razvanch) [inline 01-basics/src/Complex.java]: It's not very clear how you want this to be implemented. With 're' and 'im' declared as private, you might want to mention that getters / setters should be also added to this class.

File: 01-basics/src/ComplexOperations.java [hunk 35]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Created by nfswo on 26-Mar-17.
+ */
+public class ComplexOperations {
+
+    // adds two complex numbers, return a new Complex one
+    public Complex addTwoComplexNumbers(Complex c1, Complex c2){}
```
Comment (razvanch) [inline 01-basics/src/ComplexOperations.java]: space before { and move the closing } on another row
couldo also add a todo comment, '// TODO: implement complex number addition'

File: 01-basics/src/Test.java [hunk 36]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Created by nfswo on 26-Mar-17.
+ */
+public class Test {
+
+    public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/src/Test.java]: space before {

File: 01-basics/Basics.rst [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+
+Clasele sunt scheletele care contin modelarea si comportamentul
+care va fi preluat de catre instante.
+
+Instantele (numite si obiecte) sunt exact ca niste variabile care au
+comportamentul definit in clasele respective.
+
+Exemplu clasa:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+
+Constructorul este o metoda care se apeleaza la instantierea unui obiect
+(la apelarea keywordului new). Daca nu definim noi unul, compilatorul ne va
+face unul default, care initializeaza toate campurile cu valorile lor default
+(pentru int/float este 0, pentru String este NULL, etc).
+
+Pentru a defini un constructor, nu ii vom pune tip de return si acesta va
+avea numele clasei:
+
+public class Test{
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/Basics.rst [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+  public Test(int x, int y){
+    this.x = x; // setam x si y din clasa sa fie cei dati ca parametru
+    this.y = y;
+  }
+
+  public void showXY(){
+    System.out.println("x este: " + this.x + ", iar y: " + this.y); // "+" concateneaza doua stringuri
+  }
+
+  public static void main(String[] args){
```
Comment (razvanch) [inline 01-basics/Basics.rst]: space before {

File: 01-basics/src/Complex.java [hunk 40]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Created by nfswo on 25-Mar-17.
+ */
+public class Complex {
+    private float re;
+    private float im;
+
+    // constructor care seteaza campurile clasei
+    public Complex(float re, float im){}
```
Comment (razvanch) [inline 01-basics/src/Complex.java]: space before { and move the closing } on another row

File: 01-basics/src/Complex.java [hunk 41]
```diff
@@ -0,0 +1,16 @@
...
+ */
+public class Complex {
+    private float re;
+    private float im;
+
+    // constructor care seteaza campurile clasei
+    public Complex(float re, float im){}
+
+    // constructor care face campurile 0
+    public Complex() {}
```
Comment (razvanch) [inline 01-basics/src/Complex.java]: move the closing } on another row

File: 01-basics/src/Complex.java [hunk 42]
```diff
@@ -0,0 +1,16 @@
...
+
+    // afisare de genul "a + bi"
+    public String toString() {}
```
Comment (razvanch) [inline 01-basics/src/Complex.java]: move the closing } on another row
couldo also add a todo comment, '// TODO: pretty print complex number'

**Review by razvanch — CHANGES_REQUESTED** [review]
Good job but still needs a few improvements.

File: 01-basics/src/Complex.java (hunk 34 above)
Comment (Adrianzatreanu) [reply 01-basics/src/Complex.java]: true, doing so.

**Pull request merged**